File: languages/python/polar/ffi.py
from contextlib import contextmanager
from dataclasses import dataclass
import json

from _polar_lib import ffi, lib

from .errors import get_python_error
from .exceptions import PolarRuntimeException


class Polar:
    def __init__(self):
        self.ptr = lib.polar_new()

    def __del__(self):
        lib.polar_free(self.ptr)

    def new_id(self):
        """Request a unique ID from the canonical external ID tracker."""
        return check_result(lib.polar_get_external_id(self.ptr))

    def load_str(self, string, filename):
        """Load a Polar string, checking that all inline queries succeed."""
        string = to_c_str(string)
        filename = to_c_str(str(filename)) if filename else ffi.NULL
        check_result(lib.polar_load(self.ptr, string, filename))

    def new_query_from_str(self, query_str):
        return Query(
            check_result(lib.polar_new_query(self.ptr, to_c_str(query_str), 0))
        )

    def new_query_from_term(self, query_term):
        return Query(
            check_result(
                lib.polar_new_query_from_term(self.ptr, ffi_serialize(query_term), 0)
            )
        )

    def next_inline_query(self):
        q = lib.polar_next_inline_query(self.ptr, 0)
        if is_null(q):
            return None
        else:
            return Query(q)

    def register_constant(self, name, value):
        name = to_c_str(name)
        value = ffi_serialize(value)
        check_result(lib.polar_register_constant(self.ptr, name, value))


class Query:
    def __init__(self, ptr):
        self.ptr = ptr

    def __del__(self):
        lib.query_free(self.ptr)

    def call_result(self, call_id, value):
        """Make an external call and propagate FFI errors."""
        if value is None:
            value = ffi.NULL
        else:
            value = ffi_serialize(value)
        check_result(lib.polar_call_result(self.ptr, call_id, value))

    def question_result(self, call_id, answer):
        answer = 1 if answer else 0
        check_result(lib.polar_question_result(self.ptr, call_id, answer))

    def application_error(self, message):
        """Pass an error back to polar to get stack trace and other info."""
        message = to_c_str(message)
        check_result(lib.polar_application_error(self.ptr, message))

    def next_event(self):
        return QueryEvent(check_result(lib.polar_next_query_event(self.ptr)))

    def debug_command(self, command):
        check_result(lib.polar_debug_command(self.ptr, ffi_serialize(command)))


class QueryEvent:
    def __init__(self, ptr):
        self.ptr = ptr

    def get(self):
        return ffi.string(self.ptr).decode()

    def __del__(self):
        lib.string_free(self.ptr)


class Error:
    def __init__(self):
        self.ptr = lib.polar_get_error()

    def get(self):
        return get_python_error(ffi.string(self.ptr).decode())

    def __del__(self):
        lib.string_free(self.ptr)


def check_result(result):
    if result == 0 or is_null(result):
        raise Error().get()
    return result


def is_null(result):
    return result == ffi.NULL


def to_c_str(string):
    return ffi.new("char[]", string.encode())


def ffi_serialize(value):
    return to_c_str(json.dumps(value))

File: languages/java/polar/src/Ffi.java
import jnr.ffi.LibraryLoader;
import jnr.ffi.Pointer;

public class Ffi {
    private PolarLib polarLib;

    protected class PolarPtr {
        private Pointer ptr;

        protected PolarPtr(Pointer ptr) {
            this.ptr = ptr;
        }

        protected Pointer get() {
            return ptr;
        }

        @Override
        protected void finalize() {
            polarLib.polar_free(ptr);
        }
    }

    protected class QueryPtr {
        private Pointer ptr;

        protected QueryPtr(Pointer ptr) {
            this.ptr = ptr;
        }

        protected Pointer get() {
            return ptr;
        }

        protected int polarQuestionResult(long call_id, int result) throws Exceptions.OsoException {
            return checkResult(polarLib.polar_question_result(ptr, call_id, result));
        }

        protected int polarCallResult(long call_id, String value) throws Exceptions.OsoException {
            return checkResult(polarLib.polar_call_result(ptr, call_id, value));
        }

        protected String polarNextQueryEvent() throws Exceptions.OsoException {
            return checkResult(polarLib.polar_next_query_event(ptr));
        }

        @Override
        protected void finalize() {
            polarLib.polar_free(ptr);
        }

    }

    protected static interface PolarLib {
        int polar_debug_command(Pointer query_ptr, String value);

        int polar_free(Pointer polar);

        String polar_get_error();

        long polar_get_external_id(Pointer polar_ptr);

        int polar_load(Pointer polar_ptr, String src, String filename);

        Pointer polar_new();

        Pointer polar_new_query(Pointer polar_ptr, String query_str);

        Pointer polar_new_query_from_term(Pointer polar_ptr, String query_term);

        Pointer polar_next_inline_query(Pointer polar_ptr);

        String polar_next_query_event(Pointer query_ptr);

        Pointer polar_query_from_repl(Pointer polar_ptr);

        int polar_question_result(Pointer query_ptr, long call_id, int result);

        int polar_call_result(Pointer query_ptr, long call_id, String value);

        int query_free(Pointer query);

        int string_free(Pointer s);

    }

    protected Ffi() {
        polarLib = LibraryLoader.create(PolarLib.class).load("lib/libpolar.dylib");
    }

    protected int polarDebugCommand(QueryPtr queryPtr, String value) throws Exceptions.OsoException {
        return checkResult(polarLib.polar_debug_command(queryPtr.get(), value));
    }

    protected long polarGetExternalId(PolarPtr polarPtr) throws Exceptions.OsoException {
        return checkResult(polarLib.polar_get_external_id(polarPtr.get()));
    }

    protected int polarLoad(PolarPtr polarPtr, String src, String filename) throws Exceptions.OsoException {
        return checkResult(polarLib.polar_load(polarPtr.get(), src, filename));
    }

    protected PolarPtr polarNew() throws Exceptions.OsoException {
        return new PolarPtr(checkResult(polarLib.polar_new()));
    }

    protected QueryPtr polarNewQuery(PolarPtr polarPtr, String queryStr) throws Exceptions.OsoException {
        return new QueryPtr(checkResult(polarLib.polar_new_query(polarPtr.get(), queryStr)));
    }

    protected QueryPtr polarNewQueryFromTerm(PolarPtr polarPtr, String queryTerm) throws Exceptions.OsoException {
        return new QueryPtr(checkResult(polarLib.polar_new_query_from_term(polarPtr.get(), queryTerm)));
    }

    protected QueryPtr polarNextInlineQuery(PolarPtr polarPtr) throws Exceptions.OsoException {
        // Don't check result here because the returned Pointer is null to indicate
        // termination
        Pointer p = polarLib.polar_next_inline_query(polarPtr.get());
        if (p == null) {
            return null;
        } else {
            return new QueryPtr(p);
        }
    }

    protected QueryPtr polarQueryFromRepl(PolarPtr polarPtr) throws Exceptions.OsoException {
        return new QueryPtr(checkResult(polarLib.polar_query_from_repl(polarPtr.get())));
    }

    protected int stringFree(Pointer s) throws Exceptions.OsoException {
        return checkResult(polarLib.string_free(s));
    }

    private Exceptions.OsoException getError() {
        return Exceptions.getJavaError(polarLib.polar_get_error());
    }

    private int checkResult(int i) throws Exceptions.OsoException {
        if (i == 0) {
            throw getError();
        } else {
            return i;
        }
    }

    private long checkResult(long i) throws Exceptions.OsoException {
        if (i == 0) {
            throw getError();
        } else {
            return i;
        }
    }

    private Pointer checkResult(Pointer p) throws Exceptions.OsoException {
        if (p == null) {
            throw getError();
        } else {
            return p;
        }
    }

    private String checkResult(String s) throws Exceptions.OsoException {
        if (s == null) {
            throw getError();
        } else {
            return s;
        }

    }
}
